Add id to users with same name

Sushmey opened the pull request:
<!-- Please see https://github.com/zulip/zulip-terminal#contributor-guidelines ! -->

**What does this PR do?**  <!-- Overall description goes here -->
On the Zulip terminal, users with same name weren't discernable because they were considered the same users. This PR compares users based on their user id and then displays the user id beside their user name if they have the same user name.
<!-- If fixing a filed bug or new feature, add 'Fixes #<issue>' or 'Partial fix for #<issue>' -->
Fixes #1151 
<!-- Add a link to a discussion on chat.zulip.org, if relevant -->

**Tested?** <!-- Fine to leave some of these unchecked if this is a draft/work-in-progress -->
- [x] Manually
- [x] Existing tests (adapted, if necessary)
- [x] New tests added (for any new behavior)
- [x] Passed linting & tests (each commit)
<!-- Code must pass CI (GitHub Actions) before merging - look for the green tick! -->

<!-- See https://github.com/zulip/zulip-terminal#commit-style -->
**Commit flow** <!-- if more than one commit; add/delete/fill-in as appropriate -->
**first commit** -> comparing users based on user_id instead of name so that a separate header can be displayed


**second commit** -> displaying user_id beside all the users who have same name as more than one user in the list of all users. Adding a new style class so that the appended user_id doesn't look like a part of the username


**third commit** -> changing label from `names` to `msg_sender` in all colour themes
<!-- For example:
- first commit doing some thing
- maybe multiple commits doing similar things
-->

**Notes & Questions** <!-- if any; add/delete/fill-in as appropriate -->
This makes users with same name have different headers by appending the id to the author name and shows it as a visually separate addition.
<!-- For example:
- this doesn't include feature X (yet?)
- unsure about Y
- should this do Z?
-->

**Interactions** <!-- if any; add/delete/fill-in as appropriate -->
- Waiting on PR review 
<!-- eg.
- Waiting on PR review 
- Blocks #<PR>
-->

**Visual changes** <!-- if any; add/delete/fill-in with screenshot/diagram as appropriate -->
Before:
<img width="1055" alt="Screenshot 2022-09-13 at 7 44 25 PM" src="https://user-images.githubusercontent.com/71403193/190386027-0ca765df-39e2-4e72-8552-3893fdb0c747.png">

After:
<img width="426" alt="Screenshot 2022-10-21 at 11 48 24 AM" src="https://user-images.githubusercontent.com/71403193/197126388-5d3bdf0c-1ab8-4a48-ba12-89c488cd2725.png">

Sushmey committed:
bugfix: messages: Compare users based on user_id.

Users with same name had their adjacent messages displayed under the same
header and no visually discerning feature was available for users with same
name. This solves that.

Fixes #1151.

mounilKshah commented on zulipterminal/model.py:
```diff
@@ -1144,6 +1147,14 @@ def user_name_from_id(self, user_id: int) -> str:
+        """
+        Returns the count of the users with the same name as the given name.
+        """
+        if not self.user_name_dict_count[user_name]:
+            return -1
+        return self.user_name_dict_count[user_name]
```
I think it would be better/safer if you used `.get()` to access the count, instead of using the index.

mounilKshah commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -76,7 +76,12 @@ def __init__(self, view: Any) -> None:
         self.model = view.model
         self.view = view
-
+        # global users_list  # Making it global so it can be accessed in the MessageBox class
+        # users_list = self.view.users
+        # global user_names_counter  # Making it global so it needs to be calculated only once
+        # user_names_counter = Counter(
+        #     user["full_name"] for user in users_list
+        # )  # Counting number of users having same name
```
This commented code seems redundant. If you don't think it is necessary to the PR, then you may remove it.

mounilKshah reviewed (changes requested):
I ran your code manually and it worked fine for me. I tried going through the files to give an in-depth review, but it is difficult to review due to the current commit structure. It would be great if you could squash the commits using `git rebase`.
Also, upon running `pytest` on your PR, there were errors. All those can be looked into deeper once the commit structure is streamlined.

neiljp commented:
@Sushmey I just merged the last commit you had and merged that first with slight adjustments, since it does the independent rename of name->msg_sender.

Some things to note:
- this change was quite separate, so having it first in the sequence of commits would mean it'd have been closer to the original code branch (now main), so easier to partially merge first if it's a small preparation for others and fits in the same PR
- since it was the last commit, I adjusted the UI code to work without the changes - not much effort this time :)
- I updated the commit title & description:
  - in ZT we include the file names as an 'area' if they can be made to fit
  - generally it's good to use the present tense, so 'Rename' here vs 'Renaming'
  - this particular commit doesn't fix 1151 :)
  - it's useful to put the 'why' in the commit description

I'll take a look at the other commits.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -1566,12 +1565,19 @@ def main_view(self) -> List[Any]:
-
-            if any(different[key] for key in ("recipients", "author", "24h")):
-                text["author"] = ("name", message["this"]["author"])
+            TextType = Dict[str, List[urwid_MarkupTuple]]
```
This change, with accompanying adjustments, is a perfectly fine separate refactor commit.

Generally more commits are better than fewer, since they're more difficult to separate later. However, note that these are ideally structured commits rather than commits that one does from just adding fixes to previous code in a branch.

Sushmey replied:
So this wasn't purely a refactor since adding a new style class to the urwid Text widget needed me to change the TextType also so that it would pass all `make check` checks. 

But I understand what you meant by keeping commits structured, hopefully the newer ones I made are more suitable

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -2221,11 +2224,11 @@ def test_msg_generates_search_and_header_bar(
         [
             (
-                [STATUS_INACTIVE, "alice", " ", "DAYDATETIME"],
+                [STATUS_INACTIVE, "alice(4)", " ", "DAYDATETIME"],
```
I think we'll probably want a space between the name and user id.

Sushmey replied:
Okay, making that change in both test cases, and boxes.py

neiljp reviewed (commented):
@Sushmey This looks like it works well :+1:

I hope you can see after the refactor merge I did, that it's possible to do partial merges step by step of larger PRs. So we can do the same with chunks of this PR that do the first stage and then the second. As I mentioned some time ago in the stream, ideally they'd have tests for the new/changed functionality, if you feel comfortable with that. Also each stage (previously just two commits) can be split into smaller changes as separate commits, notably the second stage where you add the function to the model, and also you refactor the TextType, then you can add the changes which enact the feature after that.

While I'd normally have been glad to see the commit titles and description tidied (see previous comment), it was useful to see here where you'd presumably squashed commits, since some of that is fine as separate commits to explain the process - what I think of as the 'story' of the sequence of commits making up the PR.

Note that each commit should pass `make check` (running linting and pytest), not just the overall result, and `tools/check-branch` is useful for ensuring that applies for all commits in a branch.

Sushmey commented on zulipterminal/model.py:
```diff
@@ -1130,7 +1131,9 @@ def get_all_users(self) -> List[Dict[str, Any]]:
         self.user_dict[current_user["email"]] = current_user
         self.user_id_email_dict[self.user_id] = current_user["email"]
-
+        self.user_name_dict_count = Counter(
```
Okay, refactoring name to `user_name_counter` instead of `user_name_dict_count`

Sushmey commented on zulipterminal/model.py:
```diff
@@ -1144,6 +1147,12 @@ def user_name_from_id(self, user_id: int) -> str:
+        """
+        Returns if the count of the users with the same name is more than 1.
+        """
+        return True if self.user_name_dict_count.get(user_name, -1) > 1 else False
```
Sorry missed changing the return type to bool. No, mypy didn't note this. 
Refactoring the name to `is_user_name_duplicate`

Yes, I don't need inline condition, thank you!

Sushmey commented on zulipterminal/model.py:
```diff
@@ -1018,6 +1018,7 @@ def get_all_users(self) -> List[Dict[str, Any]]:
         # and a user-id to email mapping
         self.user_dict: Dict[str, Dict[str, Any]] = dict()
         self.user_id_email_dict: Dict[int, str] = dict()
+        self.user_name_dict_count: Dict[str, int] = dict()
```
Yes, initialized it as `Counter()` instead of `dict()`

Sushmey commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -1550,7 +1548,8 @@ def main_view(self) -> List[Any]:
             "recipients": recipient_header is not None,
-            "author": message["this"]["author_id"] != message["last"]["author_id"],
+            "author": message["this"]["author"] != message["last"]["author"],
+            "author_id": message["this"]["author_id"] != message["last"]["author_id"],
```
Nope, both are not needed. Will remove it in a separate commit since I might to change test cases since it's failing some assertion test

neiljp reviewed (commented):
@Sushmey Thanks for your work on this! Aside from some minor suggestions, the test you added looks fine 👍 After restructuring, you may see other opportunities for where more/adjusted tests may make sense (see below).

Your intentions in the PR would be easier to read if you could group code into commits differently, ie. as per https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-discipline

One example, to prepare for the second change, would be to group the new method you introduced (`is_user_name_duplicate`), as well as the related model data & setup, and the test you have in a separate commit right now.

Similarly keep only the TextType changes in one commit.

Sushmey commented on tests/ui/test_ui_tools.py:
```diff
@@ -2224,11 +2224,11 @@ def test_msg_generates_search_and_header_bar(
             ),
             ([" ", " ", " ", "DAYDATETIME"], {"timestamp": 1532103779}),
-            ([STATUS_INACTIVE, "alice(4)", " ", "DAYDATETIME"], {"timestamp": 0}),
+            ([STATUS_INACTIVE, "alice (4)", " ", "DAYDATETIME"], {"timestamp": 0}),
```
I had thought at that time that it made sense for them to be in a different commit with other little refactors but now I've squashed it with other previous commits

Sushmey commented on zulipterminal/model.py:
```diff
@@ -1019,7 +1019,7 @@ def get_all_users(self) -> List[Dict[str, Any]]:
         self.user_dict: Dict[str, Dict[str, Any]] = dict()
         self.user_id_email_dict: Dict[int, str] = dict()
-        self.user_name_dict_count: Dict[str, int] = dict()
+        self.user_name_counter: Dict[str, int] = Counter()
```
Yep, fair. Done!

Sushmey commented on tests/model/test_model.py:
```diff
@@ -3774,6 +3774,21 @@ def test_user_name_from_id_invalid(self, model, user_id):
+        [
+            ("Alice", {"Greg": 2}, False),
+            ("Alice", {"Alice": 1}, False),
+            ("Greg", {"Greg": 2}, True),
```
I didn't know this. Thank you for pointing it out to me. I've added the cases with relevant messages.

Sushmey committed:
messages: Add style class to user id and update tests.

Add a style class to user id in header to differentiate from the username.

Users with same usernames have their user id appended to it.

Doing it in the same header test needs changing of all the parameters.

Hence adding a new test function specifically for that case.

mounilKshah reviewed (commented):
Thank you for your effort @Sushmey!
I tried the code and it worked pretty well for me. The commit structure looks much better than the last time I saw it. Only thing that I feel might need to be changed is the commit messages as they do not follow the [gitlint rules](https://github.com/mounilKshah/zulip-terminal#gitlint). Except for that, everything looks good to me.

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -2321,7 +2324,7 @@ def test_main_view_generates_EDITED_label(
         "to_vary_in_last_message, update_required",
         [
-            ({"sender_full_name": "Unique name (won't be in next message)"}, True),
+            ({"sender_id": "Unique id (won't be in next message)"}, True),
```
Why is this updated? Does this behavior need updating too, as a follow-up?

Sushmey replied:
The function is merely checking update of author status so might not need that change. At that time I thought it'd need it since we were shifting from comparing based on `sender_full_name` to `sender_id`. Changing it back now

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -2021,6 +2022,7 @@ def test_main_view_renders_slash_me(self, mocker, message, content, is_me_messag
         [
             {
                 "id": 4,
+                "sender_id": 4,
```
Nor here.

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -1983,6 +1983,7 @@ def test_main_view(self, mocker, message, last_message):
         [
             {
                 "id": 4,
+                "sender_id": 4,
```
This does not appear to need this update?

I understand the desire to keep this message data up to date. We should likely be using something like a `message_fixture` from `conftest.py`. These tests were initially copy and pasted quite a lot, so a good refactor in a different commit/PR would be to simplify these by using our standard fixtures. One possible point to check is whether these are fixtures representing messages we receive from the server, and whether we update them later in the model (eg. modernize message response).

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -1567,7 +1569,9 @@ def main_view(self) -> List[Any]:
             text: TextType = {key: (None, " ") for key in text_keys}
 
-            if any(different[key] for key in ("recipients", "author", "24h")):
+            if any(
+                different[key] for key in ("recipients", "author", "24h", "author_id")
+            ):
```
This is now new behavior, so while it passes existing tests, it would be good to extend or add a test to illustrate this, and include it in this commit.

neiljp reviewed (commented):
@Sushmey Thanks for your work on this - I'm trying to catch up on PRs!

This was an early PR, and unfortunately before I extracted the message box code into a separate file, so I've just migrated your changes across while rebasing. That should have been done correctly, but please check that the diffs in your local branch and the version I just pushed looks OK, even if they are in different files in some cases.

In doing the migration, I also:
- split the model changes out into a separate commit from the UI update
- updated the commit text a little

Some of my inline comments may not migrate cleanly since I'm pushing different files, so they may not appear in the commit-by-commit view, but will in the main conversation page.


The code definitely looks reasonable, but as per my inline comments, this would benefit from some work on the tests, including adding specific tests for the new behavior separately.

neiljp commented:
@Sushmey Ah, I'm also finding the tests are failing locally, so that may have been a migration issue on my part.

I also meant to say that the last commit doesn't have test changes and clearly changes behavior, which is one reason I'd like to see a new test for previous commit, so that the code change can match a test change.

Sushmey commented on tests/ui/test_ui_tools.py:
```diff
@@ -2221,11 +2224,11 @@ def test_msg_generates_search_and_header_bar(
             ),
             ([" ", " ", " ", "DAYDATETIME"], {"timestamp": 1532103779}),
-            ([STATUS_INACTIVE, "alice", " ", "DAYDATETIME"], {"timestamp": 0}),
+            ([STATUS_INACTIVE, "alice (4)", " ", "DAYDATETIME"], {"timestamp": 0}),
```
So I tried changing it back to normal (i.e. name without user id) but it was still passing the duplicate name condition and appending `None`. At first I thought the test user_list has `alice` as a common name but changing it to a unique name also gave the same results. Then I realised the `this_msg` and `last_msg` in the test function are derived from the same parameters and they have the same sender name (although different `sender_id`) and that's why its always appending the `sender_id` to the name.

Sushmey replied:
I looked into this a bit more and realised there's no way to include the duplicate username case in the earlier function where we test headers  without appending the userid to every name in the parameters because we are keeping the same message and varying just a few attributes. Also, added a mock response as `false` for the `is_user_name_duplicate` function so that it doesn't fail for the reasons mentioned earlier, and since duplicate usernames aren't what we are testing in that test anyway.

Finally, I decided to make a new test for testing the specific case where username is a duplicate where it would need userid appended to it.

Sushmey committed:
model: Track full name count, enabling new is_user_name duplicate().

Added a function that checks if username is duplicate.

Added tests for the same.

Sushmey commented:
@neiljp Thank you so much for the review! 🤗  

I've made changes based on your comments. 

I added a separate test for duplicate usernames but different author ids to account for the new behaviour and updated a test case for the same.  I tried including it in the same test but it doesn't seem possible without having to change ever test case parameter. Also, added a mock response as false for the is_user_name_duplicate function so that it doesn't fail for the reasons mentioned earlier, and since duplicate usernames aren't what we are testing in that test anyway.
